[Kang Wenhan] iP

onepersonhere opened the pull request:
# Renaissance Duke

> All new original Duke with a twist 

Renaissance Duke is a very **perculiar** chat bot that helps you keep track of your daily task! 🙃

Features:
- Differation of Task
- Search by date
- Search by Keyword
- Load & Save 
- And… more!

Commands:
1. date (search by date)
2. find (search by keyword)
3. deadline (add a dateline)
4. event (add a event)
5. todo (add a todo)
6. delete (delete a task)
7. list (list all the current task)
8. load (loads a task from file)
9. save (saves a task into file)
10. mark (marks the task as done)
11. unmark (unmarks the task as done)
12. exit (exits Duke) 

Download it [here](https://github.com/onepersonhere/ip/releases/tag/A-Jar)!

Future plans:

 - [x]  Managing more tasks
 - [ ] Reminder
 - [ ] Multiple language support
 - [ ] Voiceover
__________

To add more `Commands` to Renaissance Duke, do extend this `abstract` class:
```java
 public abstract class Command {
	protected boolean isExit = false;

    /**
     * Sets the exit flag.
     *
     * @return true if the exit flag is set, false otherwise.
     */
    public boolean isExit() {
        return isExit;
    }

    /**
     * Executes the command.
     *
     * @param ui the user interface
     * @param storageList the storage list
     * @throws DukeException if an error occurs
     */
    public abstract void execute(Ui ui, StorageList storageList) throws DukeException;
}
```

Commit:
Add Gradle support

onepersonhere committed:
Implemented a skeletal version of Duke that starts by greeting the user, simply echos commands entered by the user, and exits when the user types bye.

onepersonhere committed:
Add the ability to store whatever text entered by the user and display them back to the user when requested.

onepersonhere committed:
Added the ability to mark tasks as done. Optionally, add the ability to change the status back to not done.

onepersonhere committed:
Added support for tracking three types of tasks

onepersonhere committed:
Semi-automated the testing of Duke

onepersonhere committed:
Dealt with errors such as incorrect inputs entered by the user.

onepersonhere committed:
Added support for deleting tasks from the list.

onepersonhere committed:
Added dates and times

Stored deadline dates as a LocalDateTime in task objects.

Accept dates in a format such as yyyy-mm-dd HHmm format (e.g., 2019-10-15 1800) and print in a different format such as HH:mm, E, MMM dd yyyy e.g., (18:00, tues, Oct 15 2019).

onepersonhere committed:
Add autosave and autoload capabilities

Save the tasks in the hard disk automatically whenever the task list changes. Load the data from the hard disk when Duke starts up.

onepersonhere committed:
Merge branch 'branch-Level-7'

onepersonhere committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/Duke.java
#	src/main/java/Output.java
#	src/main/java/StorageList.java
#	text-ui-test/EXPECTED.TXT
#	text-ui-test/input.txt

onepersonhere committed:
Make the code more OOP

Refactor the code to extract out closely related code as classes.
Ui: deals with interactions with the user.
FileIO: deals with loading tasks from the file and saving tasks in the file.
Parser: deals with making sense of the user command.
StorageList: contains the task list e.g., it has operations to add/delete tasks in the list.
Command: contains the abstract logic for *Command classes that inheirits from it.

By refactoring the code to be more OOP, we are able to model
standard working modules to communicate with one another, this in turns allows us to
break the program into the bit-sized problems that can be solved easily. This will give us
greater programmer productivity, better quality of software and lesser maintenance cost.
This will also means that we can eliminate redundant code and extend the use of existing classes,
by using inheritance. However, this will definitely increase the size of the code base.

onepersonhere committed:
Organize into Packages

Organize the classes into suitable java packages.

Classes are divided into the following:
1. duke.task: holds tasks
2. duke.command: holds commands
3. duke: holds main files

onepersonhere committed:
Merge pull request #1 from nus-cs2103-AY2223S1/add-gradle-support

Add Gradle support

onepersonhere committed:
Automate project builds using Gradle

Gradle is setup to run and build Duke and run unit tests.

onepersonhere committed:
Add JUnit tests

Add JUnit tests to test the behavior of the code.

Tests added:
AddTest
MarkTest
ListTest

onepersonhere committed:
Change the archiveBaseName

archiveBaseName = Duke
archiveVersion = 0.1

onepersonhere committed:
Add JavaDoc comments

Add JavaDoc comments to the code.

Add header comments to all non-private classes/methods, and non-trivial private methods.

onepersonhere committed:
Tweak the code to comply with a coding standard

Tweak the code to comply with the given coding standard.
From this point onward, any new code added are compliant with the given coding standard.

onepersonhere committed:
Enables find tasks by keyword

Give users a way to find a task by searching for a keyword.

onepersonhere committed:
Merge branch 'branch-A-JavaDoc'

onepersonhere committed:
Merge branch 'branch-Level-9'

onepersonhere committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/duke/Duke.java
#	src/main/java/duke/DukeException.java
#	src/main/java/duke/FileIO.java
#	src/main/java/duke/Output.java
#	src/main/java/duke/Parser.java
#	src/main/java/duke/StorageList.java
#	src/main/java/duke/Ui.java
#	src/main/java/duke/command/AddCommand.java
#	src/main/java/duke/command/Command.java
#	src/main/java/duke/command/DeleteCommand.java
#	src/main/java/duke/command/ExitCommand.java
#	src/main/java/duke/command/ListCommand.java
#	src/main/java/duke/command/LoadCommand.java
#	src/main/java/duke/command/MarkCommand.java
#	src/main/java/duke/command/MatchCommand.java
#	src/main/java/duke/command/SaveCommand.java
#	src/main/java/duke/command/UnmarkCommand.java
#	src/main/java/duke/task/Deadline.java
#	src/main/java/duke/task/Event.java
#	src/main/java/duke/task/Task.java
#	src/main/java/duke/task/Todo.java

onepersonhere committed:
Update README.md

onepersonhere committed:
Add a GUI to Duke

Used the JavaFX technology to implement the GUI.

onepersonhere committed:
Merge branch 'branch-Level-10'

ekweirui commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,138 @@
+     * @param storageList StorageList object
+     * @param secCommand  Second command that the user inputs
+     */
+    public static void parseTask(String input, String command, StorageList storageList, String secCommand) {
+        try {
+            String desc = findFirstCommand(input, command);
+            String secondCommand = findSecondCommand(input, secCommand);
+            switch (command) {
+            case "deadline":
+                addTask(new Deadline(desc, secondCommand), storageList);
+                break;
+            case "event":
+                addTask(new Event(desc, secondCommand), storageList);
+                break;
+            case "todo":
+                addTask(new Todo(desc), storageList);
+                break;
+            default:
+            }
+            FileIO.save(storageList, "./default.txt");
+        } catch (DukeException e) {
+            System.out.println(e.getMessage());
+        }
+    }
```
Perhaps this block of code would be better placed in the three addcommand classes, as it might be weird for the other classes to be handled individually but addcommand handled here.

onepersonhere replied:
That is true, however this class is suppose to be able to parse and make sense of the command the user has type, which I why I feel that putting this function in Parser class is more approprite.

ekweirui commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,138 @@
+     * @param task Task to be added to the StorageList
+     * @param storageList StorageList to be added to
+     */
+    private static void addTask(Task task, StorageList storageList) {
+        storageList.add(task);
+        Output.ADD.modifyTask(task, storageList);
+    }
```
Perhaps this piece of code would be better in the storageList class as it is a behaviour of a storageList and not parser itself.

onepersonhere replied:
Thanks for the review, this class will be moved to StorageList.

ekweirui commented on src/main/java/duke/StorageList.java:
```diff
@@ -0,0 +1,152 @@
+        list.clear();
+    }
+
+    /**
```
I like that you implemented several ways to filter the tasks!

onepersonhere replied:
Thanks!

ekweirui commented on src/main/java/duke/command/AddCommand.java:
```diff
@@ -0,0 +1,39 @@
+ * Add a new Task to the StorageList.
+ */
+public class AddCommand extends Command {
+    /**
+     * Executes the add command.
+     *
+     * @param ui the user interface
+     * @param storageList the storage list
+     * @throws DukeException if the command is invalid
+     */
```
I feel that it would be tidier and more efficient to put Javadoc comment for just the overridden method in the abstract command class as it can potentially save a lot of your time. (javadoc comments still applies for overriding method)

onepersonhere replied:
Thanks, will do so for future overriden methods!

ekweirui reviewed (commented):
Overall, your code looks really good and clean! I admire how you gave more than the specified ways to filter out tasks and print them out. I learnt alot from reading your code and I think it is definitely on the right track. Other than a few small suggestions, I think you did a great job! 😃

senwei01 commented on src/main/java/duke/gui/MainWindow.java:
```diff
@@ -0,0 +1,85 @@
+    }
+
+    /**
+     * Sets the duke object.
```
From coding standard, Javadoc comments require a newline between description and param section. I've only spotted this here, in general you followed the coding standard pretty well

onepersonhere replied:
Thanks for the spot!

senwei01 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,138 @@
+     * @param task Task to be added to the StorageList
+     * @param storageList StorageList to be added to
+     */
+    private static void addTask(Task task, StorageList storageList) {
```
Should this be in StorageList instead? Might want to consider it as it is more of a function of storageList.

onepersonhere replied:
Thanks for the review, this class will be moved to StorageList.

senwei01 commented on src/main/java/duke/Output.java:
```diff
@@ -0,0 +1,135 @@
+package duke;
```
I like how you have a class just for the output :)

onepersonhere replied:
Thanks!

senwei01 reviewed (commented):
Overall I really like the readability of your code. I think you have great OOP as well!

onepersonhere committed:
Update build.gradle

Changed mainClassName

onepersonhere committed:
Move addTask method to StorageList

onepersonhere committed:
Use assert feature

To document important assumptions that should hold at various points in
the code.

Included assert in various methods to check if the arguments are not
null.

This is so that arguments passed into the subsequent methods are valid
so they will not cause additional bugs.

Dilysss commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,112 @@
+    private StorageList storageList;
+    private Ui ui;
+    private ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
+    /**
```
Try to leave lines in between to make the code neater

Dilysss commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,126 @@
+     * @throws DukeException if the command is invalid
+     */
+    public static Command parse(String command) throws DukeException {
+        switch (command) {
```
Good that you used switch statements instead of if-else to make the code more readable

Dilysss commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,126 @@
+    public static String findFirstCommand(String input, String command) throws DukeException {
+        int endOfCommand = input.indexOf("/");
+        int beginIndex = input.indexOf(command) + command.length() + 1;
+        if (beginIndex > input.length()) {
```
```suggestion
boolean isIncorrectInput = beginIndex > input.length();
```
Similarly for line 81, where the expressions can be simplified

Dilysss reviewed (commented):
Well written code!

onepersonhere committed:
Improve code quality

Code needs to be read, understood for modification by other developers
later on.

Refactored and renamed FileIO to FileManager.
Reorganized the logic of FileManager into more methods.

This is to follow to SLAP principle by avoiding varying the level of
abstraction within a code fragment.

onepersonhere committed:
Merge pull request #2 from onepersonhere/branch-A-CodeQuality

Improve code quality

onepersonhere committed:
Merge pull request #3 from onepersonhere/branch-A-Assertions

Use assert feature

onepersonhere committed:
Make the command syntax more flexible

Long commands takes time to type.

Provided shorter aliases for keywords.

Shorter commands are faster to type.

onepersonhere committed:
Merge pull request #4 from onepersonhere/branch-C-FriendlierSyntax

Make the command syntax more flexible

onepersonhere committed:
Correct reuse credits in javadocs

Changed the reused credit style to follow the required convention

onepersonhere committed:
Add a User Guide

Included as much information as necessary for the user,
in a format as friendly to the user as possible.

onepersonhere committed:
Correct JavaDocs comments

Comments changed to singular form for the first word.

onepersonhere committed:
Correct code quality

Some header comments are modified according to code quality guidelines.
